### PR TITLE
Add Marshal ansi conversion to shared partition

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems [hunk 1]
```diff
@@ -993,6 +993,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\IO\DisableMediaInsertionPrompt.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\MemoryFailPoint.Windows.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\InteropServices\Marshal.Windows.cs" Condition="'$(TargetsCoreRT)' != 'true'" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\InteropServices\Marshal.StringToAnsiString.cs" Condition="'$(TargetsCoreRT)' != 'true'" />
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems]: Can this be part of in `Marshal.Windows.cs`, similar to Unix ?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems]: WideCharToMultiByte in the Unix Win32 PAL ignores `WC_NO_BEST_FIT_CHARS` and it always returns false for `usedDefaultChar`.

https://github.com/dotnet/coreclr/blob/master/src/pal/src/locale/unicode.cpp#L629

File: src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs [hunk 2]
```diff
@@ -2,10 +2,25 @@
...
         private static bool IsWin32Atom(IntPtr ptr) => false;
+
+        internal static unsafe int StringToAnsiString(string s, byte* buffer, int wideCharLen, bool bestFit = false, bool throwOnUnmappableChar = false)
+        {
+            int convertedBytes;
+
+            fixed (char* pChar = s)
+            {
+                convertedBytes = Encoding.UTF8.GetBytes(pChar, s.Length, buffer, wideCharLen);
+                buffer[convertedBytes] = 0;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: Is buffer guaranteed to be at least convertedBytes+1 long?

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: (This can also be moved outside of the fixed.)

Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: >Is buffer guaranteed to be at least convertedBytes+1 long?

Yes.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: Can we pass in the buffer length and assert it?

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: It looks like we already assert it in the Windows code; we should do the same in the Unix code.

Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: ok, copied the same assert from windows version

**Comment by jkotas** [comment]
`shared/System/Runtime/InteropServices/Marshal.Unix.cs(15,13): error CS0103: The name 'Debug' does not exist in the current context`

File: src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs (hunk 2 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.Unix.cs]: Could you please also rename the `wideCharLen` argument to better capture the meaning, e.g. `bufferLength` ?

**Pull request merged**